### PR TITLE
OCPBUGS-19391: do not reconcile rules if aggregationRule is set

### PR DESCRIPTION
**Pull request by PratikMahajan**
control-plane manages the cluster role rules if `aggregationRule` are set.

### CODE REVIEW

**Review by wking — APPROVED** [review]
/lgtm
/hold for pre-merge testing

**Review by petr-muller — CHANGES_REQUESTED** [review]
Fix looks good. `EnsureClusterRole` does not have unit tests, this PR seems like a good opportunity to add some? Nothing fancy, I'd say that even adding one for this new behavior (existing has aggregation rules and rules, required has just aggregation rules => no reconciliation needed) is enough.

**Comment by petr-muller** [comment]
/lgtm
I did not really want to override Trevor's review, meant just as a suggestion

**Comment by wking** [comment]
Pre-merge testing, comparing [a recent  nightly][1]:

```console
$ curl -s https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/logs/periodic-ci-openshift-release-master-ci-4.15-e2e-aws-ovn/1721927591801131008/artifacts/e2e-aws-ovn/gather-extra/artifacts/pods/openshift-cluster-version_cluster-version-operator-7446f586c-68jr4_cluster-version-operator.log | grep -o 'Updating.*diff.*' | sort | uniq -c
     21 Updating ClusterRoleBinding cluster-baremetal-operator due to diff:   &v1.ClusterRoleBinding{
     21 Updating ClusterRole machine-api-operator-ext-remediation due to diff:   &v1.ClusterRole{
     21 Updating ConfigMap openshift-machine-config-operator/kube-rbac-proxy due to diff:   &v1.ConfigMap{
```

And [this pull][2]:

```console
$ curl -s https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift_cluster-version-operator/993/pull-ci-openshift-cluster-version-operator-master-e2e-agnostic-ovn/1721583885906612224/artifacts/e2e-agnostic-ovn/gather-extra/artifacts/pods/openshift-cluster-version_cluster-version-operator-5b74775fb4-2nvqx_cluster-version-operator.log | grep -o 'Updating.*diff.*' | sort | uniq -c
     14 Updating ClusterRoleBinding cluster-baremetal-operator due to diff:   &v1.ClusterRoleBinding{
     14 Updating ConfigMap openshift-machine-config-operator/kube-rbac-proxy due to diff:   &v1.ConfigMap{
```

So hooray, the ClusterRole is no longer hotlooping.  That doesn't prove it still reconciles the resource if someone attempts to stomp over something like the `aggregationRule` property.  Seems pretty clear to me from the code that it will, but still probably worth testing via Cluster Bot or something before merging.  Petr's suggestion to test via unit tests would work too.


[1]: https://prow.ci.openshift.org/view/gs/origin-ci-test/logs/periodic-ci-openshift-release-master-ci-4.15-e2e-aws-ovn/1721927591801131008
[2]: https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_cluster-version-operator/993/pull-ci-openshift-cluster-version-operator-master-e2e-agnostic-ovn/1721583885906612224

**Comment by shellyyang1989** [comment]
Pre-merge testing 

```
# oc adm upgrade 
Cluster version is 4.15.0-0.test-2023-11-08-045808-ci-ln-48xxz5b-latest

warning: Cannot display available updates:
  Reason: NoChannel
  Message: The update channel has not been configured.
```

```
# oc logs cluster-version-operator-6f8778c7f8-jrkjh -n openshift-cluster-version | grep -o 'Updating .*due to diff'| sort | uniq -c
     40 Updating ClusterRoleBinding cluster-baremetal-operator due to diff
     40 Updating ConfigMap openshift-machine-config-operator/kube-rbac-proxy due to diff
```

No hotloop on ClusterRole machine-api-operator-ext-remediation any more.

**Comment by wking** [comment]
[I'd held for pre-merge testing][1], and [we have that now][2].

/hold cancel

[1]: https://github.com/openshift/cluster-version-operator/pull/993#pullrequestreview-1716918769
[2]: https://github.com/openshift/cluster-version-operator/pull/993#issuecomment-1801241360

**Comment by wking** [comment]
I dunno where Tide got off to,  but merging by hand because Prow is happy.

**Pull request merged**
